provider/pkg/autonaming/application.go

```diff
@@ -101,3 +135,25 @@ func getDefaultName(
 
 	return resource.NewStringProperty(random), nil
 }
+
+// trimName will trim the prefix to fit within the max length constraint.
+// It will cut out part of the middle, keeping the beginning and the end of the string.
```

Comment:
I think this is the right thing to do given the current constraints. But it could also be rather surprising to users.

For example, I often split my resource names with hyphens. This could generate some rather odd names.

But given that https://github.com/pulumi/pulumi/discussions/17592 is not ready yet, I think this is a good choice.
If we see the need for it, we could also expose more configuration options for autonaming in the future.